File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

vector<int> answer;
int sum = 0;

bool visited[100][100];
int dr[] = { 0, 0, 1, -1 };
int dc[] = { 1, -1, 0, 0 };

void BFS(vector<string> &maps) {
    for(int r = 0; r < maps.size(); r++) {
        for(int c = 0; c < maps[r].length(); c++) {
            if(maps[r][c] != 'X' && !visited[r][c]) {
                queue<pair<int, int>> q;
                
                visited[r][c] = true;
                q.push(make_pair(r, c));
                sum += maps[r][c] - '0';
                
                while(!q.empty()) {
                    int cr = q.front().first;
                    int cc = q.front().second;
                    q.pop();
                    
                    for(int i = 0; i < 4; i++){
                        int nr = cr + dr[i];
                        int nc = cc + dc[i];
                            
                        if(nr < 0 || nc < 0 || nr >= maps.size() || nc >= maps[r].length())
                            continue;
                        if(visited[nr][nc])
                            continue;
                        if(maps[nr][nc] == 'X')
                            continue;
                        
                        visited[nr][nc] = true;
                        q.push({nr, nc});
                        sum += maps[nr][nc] - '0';
                    }
                }
                
                answer.push_back(sum);
                sum = 0;
            }
        }
    }
}

vector<int> solution(vector<string> maps) {
    BFS(maps);

    if(answer.empty())
        answer.push_back(-1);
    
    sort(answer.begin(), answer.end());

    return answer;
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    let answer = [];
    
    let visited = Array.from({ length: maps.length }, () => Array(maps[0].length).fill(false));
    let dr = [ 0, 0, 1, -1 ];
    let dc = [ 1, -1, 0, 0 ];
    
    let sum = 0;
    const BFS = (r, c) => {
        let queue = [[r, c]];
        visited[r][c] = true;
        sum += parseInt(maps[r][c]);
        
        while(queue.length > 0) {
            let [cr, cc] = queue[0];
            queue.shift();
            
            for(let i = 0; i < 4; i++) {
                let [nr, nc] = [cr + dr[i], cc + dc[i]];
                
                if(nr < 0 || nc < 0 || nr >= maps.length || nc >= maps[r].length)
                    continue;
                if(visited[nr][nc])
                    continue;
                if(maps[nr][nc] === 'X')
                    continue;
                
                queue.push([nr, nc]);
                visited[nr][nc] = true;
                sum += parseInt(maps[nr][nc]);
            }
        }
        
        answer.push(sum);
        sum = 0;
    }
    
    for(let r = 0; r < maps.length; r++) {
        for(let c = 0; c < maps[r].length; c++) {
            if(maps[r][c] !== 'X' && !visited[r][c])
                BFS(r, c);
        }
    }
    
    if(!answer.length)
        answer.push(-1);
    
    answer.sort((a, b) => a - b);
    
    return answer;
}